open-apparel-registry/open-apparel-registry
# Allow list items to be matched if there are no geocode results

## Overview

The dedupe matcher may be able to match a list item to an existing, geocoded facility even if the geocoder returns no results. To allow for this we have added a new `GEOCODED_NO_RESULTS` status and modified the workflow.

Connects #275

### Demo

<img width="1222" alt="Screen Shot 2019-03-12 at 8 58 46 AM" src="https://user-images.githubusercontent.com/17363/54215458-2e625680-44a5-11e9-927a-fe773c019825.png">

<img width="1561" alt="Screen Shot 2019-03-12 at 10 02 18 AM" src="https://user-images.githubusercontent.com/17363/54220147-f9a6cd00-44ad-11e9-90f8-75ab20548299.png">

### Notes

We add a `result_count` to the response from the geocoder so we can easily distinguish a successful geocode from one that did not fail, but also did not produce any results.

Items that cannot be geocoded and also fail to be matched end up in the `ERROR_MATCHING` status because we cannot create a new Facility without a lat/lng.

While we were changing the list of available status codes we took the opportunity to add more specific error states. Creating a separate error state for each stage of processing makes the status column displayed on the facility list item page less ambiguous and prepares us for easily resubmitting items for processing at the correct point in the workflow.

## Testing Instructions

 * Verify that the tests pass
 * Apply a diff to introduce some intermittent failures
```diff
diff --git a/src/django/api/processing.py b/src/django/api/processing.py
index c791d35..23450b0 100644
--- a/src/django/api/processing.py
+++ b/src/django/api/processing.py
@@ -42,6 +42,8 @@ def parse_facility_list_item(item):
     if item.status != FacilityListItem.UPLOADED:
         raise ValueError('Items to be parsed must be in the UPLOADED status')
     try:
+        if item.id % 5 == 0:
+            raise Exception('boom')
         fields = [f.lower() for f in parse_csv_line(item.facility_list.header)]
         values = parse_csv_line(item.raw_data)
         if CsvHeaderField.COUNTRY in fields:
@@ -77,6 +79,8 @@ def geocode_facility_list_item(item):
     if item.status != FacilityListItem.PARSED:
         raise ValueError('Items to be geocoded must be in the PARSED status')
     try:
+        if item.id % 4 == 0:
+            raise Exception('boom')
         data = geocode_address(item.address, item.country_code)
         if data['result_count'] > 0:
             item.status = FacilityListItem.GEOCODED
```
 * Upload  ⁨src⁩/django⁩/api⁩/management⁩/commands⁩/facility_lists⁩/13.csv
* Parse and geocode the CSV
  * `./scripts/manage batch_process --list-id 15 --action parse`
  * `./scripts/manage batch_process --list-id 15 --action geocode`
* Verify that some items are in `ERROR_PARSING`, `ERROR_GEOCODING` and  `GEOCODED_NO_RESULTS` statuses.
* Match the CSV
   * `./scripts/manage batch_process --list-id 15 --action match`
* Verify that there is an item in the `ERROR_MATCHING` status and that it has an informative message
